sklearn/utils/extmath.py

```diff
@@ -195,6 +195,42 @@ def safe_sparse_dot(a, b, dense_output=False):
         return fast_dot(a, b)
 
 
+def _block_slices(dim_size, block_size):
+    """Generator that yields slice objects for indexing into sequential blocks
+    of an array along a particular axis
+    """
+    count = 0
+    while True:
+        yield slice(count, count + block_size, 1)
+        count += block_size
+        if count > dim_size:
+            raise StopIteration
+
+
+def _blockwise_dot(A, B, max_rows, max_cols, out=None):
+    """Computes the dot product of two matrices in a block-wise fashion. Only
+    blocks of `A` with a maximum size of `(max_rows, max_cols)` will be
+    processed simultaneously.
+    """
+    m,  n = A.shape
+    n1, o = B.shape
+    if n1 != n:
+        raise ValueError('matrices are not aligned')
+
+    if out is None:
+        out = np.empty((m, o), dtype=np.result_type(A, B))
+    elif out.shape != (m, o):
+        raise ValueError('output array has incorrect dimensions')
+
+    for mm in _block_slices(m, max_rows):
+        out[mm, :] = 0
+        for nn in _block_slices(n, max_cols):
+            A_block = A[mm, nn].copy()  # copy to force a read
```

Comment:
Why is this needed? Have you benchmarked the impact of this line? Can't the prefetch heuristics of the OS do a good job in that case? We could spare an necessary malloc here.